src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -435,26 +436,55 @@ impl ScalarFunc {
         ctx: &mut EvalContext,
         row: &'a [Datum],
     ) -> Result<Option<Cow<'a, MyDuration>>> {
-        let d0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
-        let d1: Cow<'a, MyDuration> = try_opt!(self.children[1].eval_duration(ctx, row));
-        let diff = match d0.to_nanos().checked_sub(d1.to_nanos()) {
-            Some(result) => result,
-            None => return Err(Error::overflow("DURATION", &format!("({} - {})", &d0, &d1))),
+        let arg0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let arg1: Cow<'a, MyDuration> = try_opt!(self.children[1].eval_duration(ctx, row));
+        let overflow = Error::overflow("DURATION", &format!("({} - {})", &arg0, &arg1));
+        let res = match arg0.into_owned().checked_sub(&arg1) {
+            Some(res) => res,
+            None => return Err(overflow),
         };
-        let res = MyDuration::from_nanos(diff, d0.fsp().max(d1.fsp()) as i8)?;
         Ok(Some(Cow::Owned(res)))
     }
+
+    #[inline]
+    pub fn sub_duration_and_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, MyDuration>>> {
+        let arg0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let arg1: Cow<'a, [u8]> = try_opt!(self.children[1].eval_string(ctx, row));
+        let s = ::std::str::from_utf8(&arg1)?;
+        let arg1 = MyDuration::parse(&arg1, Time::parse_fsp(s))?;
```

Comment:
It seems when agr1 is not a valid duration, should return Ok(None) in the current situtation

Reply:
Yes, you are right.

Reply:
Fixed. PTAL again.